dev/devicelab/bin/tasks/flutter_attach_test.dart

```diff
@@ -0,0 +1,154 @@
+
+  if (onListening == null)
+    listening.complete();
+
+  int exitCode;
+  process.stdout
+      .transform(utf8.decoder)
+      .transform(const LineSplitter())
+      .listen((String line) {
+    print('attach:stdout: $line');
+    stdout.add(line);
+    if (line.contains('Listening') && onListening != null) {
+      listening.complete(onListening());
+    }
+    if (line.contains('To quit, press "q".'))
+      ready.complete();
+    if (line.contains('Reloaded '))
+      reloaded.complete();
+    if (line.contains('Restarted app in '))
+      restarted.complete();
+  });
+  process.stderr
+      .transform(utf8.decoder)
+      .transform(const LineSplitter())
+      .listen((String line) {
+    print('run:stderr: $line');
+    stdout.add(line);
+  });
+
+  process.exitCode.then((int processExitCode) { exitCode = processExitCode; });
+
+  await Future.any<dynamic>(<Future<dynamic>>[ listening.future, process.exitCode ]);
+  await Future.any<dynamic>(<Future<dynamic>>[ ready.future, process.exitCode ]);
+
+  if (exitCode != null)
+    throw 'Failed to attach to test app; command unexpected exited, with exit code $exitCode.';
+
+  process.stdin.write('r');
+  process.stdin.flush();
+  await reloaded.future;
+  process.stdin.write('R');
+  process.stdin.flush();
+  await restarted.future;
+  process.stdin.write('q');
+  process.stdin.flush();
+
+  await process.exitCode;
+
+  if (stderr.isNotEmpty)
+    throw 'flutter attach had output on standard error.';
+
+  if (exitCode != 0)
+    throw 'exit code was not 0';
+
+  const List<String> expectedLinePrefixes = const <String>[
+    'Initializing hot reload...',
+    'Reloaded ',
+    'Performing hot restart...',
+    'Restarted app in ',
+    'Application finished',
+  ];
+
+  int expectedIndex = 0;
+  for (int i = 0; i < stdout.length; i++) {
+    if (stdout[i].startsWith(expectedLinePrefixes[expectedIndex])) {
+      expectedIndex++;
+      if (expectedIndex >= expectedLinePrefixes.length)
+        break;
+    }
+  }
+
+  if (expectedIndex != expectedLinePrefixes.length)
+    throw 'not all expected lines were found';
+}
+
+void main() {
+  task(() async {
+    final AndroidDevice device = await devices.workingDevice;
+    await device.unlock();
+    final Directory appDir = dir(path.join(flutterDirectory.path, 'dev/integration_tests/ui'));
+    await inDirectory(appDir, () async {
+      print('Build: starting...');
+      final String buildStdout = await eval(
+          path.join(flutterDirectory.path, 'bin', 'flutter'),
+          <String>['--suppress-analytics', 'build', 'apk', '--debug', 'lib/main.dart'],
+      );
+      final String lastLine = buildStdout.split('\n').last;
+      final RegExp builtRegExp = new RegExp(r'Built (.+)( \(|\.$)');
+      final String apkPath = builtRegExp.firstMatch(lastLine)[1];
+
+      print('Installing $apkPath');
+
+      await device.adb(<String>['install', apkPath]);
+
+      try {
+        print('Launching attach.');
```

Comment:
There is a `section` function that prints a nice heading for coarse-grained test steps.

Reply:
done.